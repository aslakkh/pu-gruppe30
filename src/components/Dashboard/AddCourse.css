.AddCourse{
    display: flex;
    flex-direction: column;
    align-items: center;
}

.AddCourseForm{
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: flex-start;
    margin-top: 10px;
    
}



