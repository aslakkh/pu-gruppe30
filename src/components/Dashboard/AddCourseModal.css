.AddCourseModal{
    width: 70%;
}

.ModalBodyWrapper{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    flex-wrap: wrap;
}

.ModalBodySeparator{
    width: 1%;
    background-color: #000;
}

.ModalBodyLeft{
    flex-basis: 45%;
}

.ModalBodyRight{
    flex-basis: 45%;
}

.ModalPane{
    background-color: darkslategray;
}

.ModalList{
    height: 400px;
    overflow-y: scroll;
}
