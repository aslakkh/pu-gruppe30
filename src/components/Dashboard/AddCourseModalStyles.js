import React from 'react';

export const styles = {
        
        listElement: {
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'flex-end',
        },
        
        listName: {
                flexBasis: '90%',
        },

        listButton: {
                marginLeft: '5px',
                marginRight: '5px',
                
        },
        modalPane: {
                backgroundColor: 'darkslategray',
        },
        
        modalList: {
                height: '400px',
                overflowY: 'scroll',
        },

        modalBodyWrapper: {
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'space-between',
                flexWrap: 'wrap',
        },

        modalBodyRight: {
                flexBasis: '45%',
        },

        modalBodyLeft: {
                flexBasis: '45%',
        },

        modalBodySeparator: {
                width: '0.2%',
                backgroundColor: '#000'
        }

}