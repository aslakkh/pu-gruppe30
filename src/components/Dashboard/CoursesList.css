.CoursesList{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.ListElementHere{
    display: flex;
    flex-direction: row;
}
