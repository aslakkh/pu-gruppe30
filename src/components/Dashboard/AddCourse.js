import React, { Component } from 'react'
import { saveCourse } from '../../helpers/auth'
import {FormGroup, FormControl, Button, HelpBlock} from 'react-bootstrap'
import './AddCourse.css'

/*      Component for adding course

Returns a form and a button.
Saves form data in firebase upon button click.
Should be updated to include better form validation. User should probably only be able to add courses that are already in database.

//https://react-bootstrap.github.io/components.html#forms

TODO

- prevent adding an already existing course from replacing previous course
- make it possible to submit form by pressing enter

*/


export default class AddCourse extends Component{
    constructor(props){
        super(props);
        this.handleAdd = this.handleAdd.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.state = {
            value: ''
        }
    }

    //get intialstate of form component. 
    getInitialState(){
        return {
            value: ''
        }
    }

    //handle change in form value. Sets the state value to form input value
    handleChange(e){
        this.setState({value: e.target.value.toUpperCase()});
        
    }

    //function for handling keypress
    handleKeyPress(e){
        if(e.key === 'Enter'){
            this.handleAdd();
        }
    }

    //function for validating input. For now returns error if form has input of length greater than 0 but less than 5, and success for anything of length > 4
    getValidationState(){
        var length = this.state.value.length;
        var re = /^[A-Z]{2,3}[0-9]{4}$/

        if(re.test(this.state.value)){
            return 'success';
        }else if(length > 0){
            return 'error';
        }
    }

    //function for handling click on add-button. Only saves course to database if input is valid.
    handleAdd = (e) => {
		e.preventDefault(); //prevents default browser behaviour on click, whatever that means

        if(this.getValidationState() === 'success'){
            saveCourse(this.state.value); //saveCourse(course) adds course to currently logged in user in Firebase
            this.setState({value: ''}); //resets value to empty string
        }
        else{
            console.log("Error: Tried submitting invalid value: " + this.state.value);
        }
		//getCourse();
	}

    //Renders a form and a button for submission
    render(){
        return(
            <div className="FormWrapper" >
                    <form>
                        <FormGroup controlId="addCourseForm" validationState={this.getValidationState()}>
                            <FormControl 
                            id="addCourseForm" 
                            componentClass='input'
                            value={this.state.value}
                            className="InputForm" 
                            type="text" 
                            placeholder="Course name"
                            onChange={this.handleChange}
                            />
                            <FormControl.Feedback/>
                            <HelpBlock>Enter course code (up to three letters followed by four digits)</HelpBlock>
                        </FormGroup>
                    </form>
                    <Button className="AddCourse-addBtn "type='submit' bsStyle="primary" bsSize="large" onClick={this.handleAdd} style={{marginLeft: "20px"}}>Add</Button>
            </div>
        )
    } 


}