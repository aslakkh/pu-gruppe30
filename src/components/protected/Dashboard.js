import React, { Component } from 'react'
import Course from './Course'
import AddCourse from './AddCourse'
import {Tab,NavItem, Nav, Col,Row} from 'react-bootstrap'
import {getCourse } from '../../helpers/auth'
export default class Dashboard extends Component {

  constructor(props){
    super();
    this.getCoursesDisp = this.getCoursesDisp.bind(this);
  }

getCoursesDisp(){
	const cour = getCourse();
	console.log(cour);

}



  render () {
    return (
    	<div>
       	<AddCourse />
      </div>
    )
  }
}