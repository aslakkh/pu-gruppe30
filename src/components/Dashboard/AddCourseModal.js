import React, {Component} from 'react'
import { saveCourse } from '../../helpers/auth'
import {Modal, Panel, FormGroup, FormControl, Button, HelpBlock, ListGroup, ListGroupItem} from 'react-bootstrap'
import firebase from 'firebase'
import AddCourse from './AddCourse'
import {HelpButton} from '../various/HelpButton'
import {styles} from './AddCourseModalStyles.js'



export default class AddCourseModal extends Component{
    constructor(props){
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.close = this.close.bind(this);
        this.state = {
            showModal: this.props.showModal,
            courses: [],
            allCourses: [],
            value: '',
        }
    }

    componentWillReceiveProps(nextProps){
        this.setState({
            showModal: nextProps.showModal,
        })
    }

    componentDidMount(){
        let that = this;
        let coursesRef = firebase.database().ref().child('courses/');
        coursesRef.on('value', snap => {
            var temp = snap.val();
            var coursesArray = Object.keys(temp).map(key => key); //create array from key values from database
            that.setState({
                allCourses: coursesArray,
                courses: coursesArray,
            });
        });
    }

    //required to use Modal. uses function closeModal received through props
    close(){
        this.props.closeModal();
        //this.setState({ showModal: false });
    }

    handleAddToMyCourses(course,e){
        e.preventDefault();
        //add to users courses in firebase
        saveCourse(course);
        /*this.setState({
                    displayFeedbackMessage: true,
                    feedbackMessage: course + " has been added to your courses!",
                    feedbackMessagePositive: true,
        });*/
    }

    handleChange(e){
        // this.setState({value: e.target.value.toUpperCase()});
        var text = e.target.value.toUpperCase();
        if(text.trim() === ""){ //if input form text is empty, set courses to all courses in database
            this.setState({
                value: text,
                courses: this.state.allCourses,
            });
        }
        else{ //use array filter to find courses matching input form text
            this.setState({
                value: text,
                courses: this.state.allCourses.filter(this.searchCourses(text)),
            });
        }
        
    }

    //used with array.filter() in handleChange-method
    //returns filter function to allow passing of parameter to filter function
    searchCourses(text){
        return function filter(value){
            return (value.search(text) !== -1);
        }
    }


    render(){
        return(
            <Modal 
            show={this.state.showModal} 
            onHide={this.close}
            dialogClassName="AddCourseModal"
            >
                <Modal.Header closeButton >
                    <Modal.Title>Add course</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <div style={styles.modalBodyWrapper}>
                        <div style={styles.modalBodyLeft} >
                            <h4>Add existing course 
                                <HelpButton
                                    placement="right"
                                    title="Adding existing courses" 
                                    helpText="Courses appearing here have been added to the database by a professor. Find your course and add it to your courses to start tracking!"
                                />
                            </h4>
                            <FormGroup controlId="addCourseForm">
                                <FormControl 
                                id="addCourseForm" 
                                componentClass='input'
                                value={this.state.value}
                                className="InputForm" 
                                type="text" 
                                placeholder="Search"
                                onChange={this.handleChange}
                                onKeyPress={this.handleKeyPress}
                                
                                />
                                <FormControl.Feedback/>
                                <HelpBlock>Enter course code (up to three letters followed by four digits)</HelpBlock>
                            </FormGroup>
                            <Panel style={styles.modalPane}>
                                <ListGroup style={styles.modalList}>
                                    {(this.state.courses).map((key) => {
                                        return <ListGroupItem key={key} style={styles.listElement}>
                                            <div style={styles.listName}>{key}</div>
                                            <Button style={styles.listButton} bsStyle="success" onClick={(e) => this.handleAddToMyCourses(key, e)}>Add to My Courses</Button>
                                        </ListGroupItem>
                                    })}
                    
                                </ListGroup>
                            </Panel>

                        </div>
                        <div style={styles.modalBodySeparator} />
                        <div style={styles.modalBodyRight}>
                            <h4>Add new course
                                <HelpButton 
                                    placement="right"
                                    title="Adding new courses"
                                    helpText="If you want to track your progress in a course that has not been added to the database by the course staff, you can enter the course code here."
                                />
                            </h4>
                            <AddCourse/>
                        </div>
                    </div>
                </Modal.Body>
            </Modal>
        )
        
    }
}