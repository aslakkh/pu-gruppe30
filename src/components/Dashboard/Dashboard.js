import React, { Component } from 'react'
import CoursesList from './CoursesList'
import AddCourseModal from './AddCourseModal'
import {Button} from 'react-bootstrap'
import firebase from 'firebase';
import {getUserUid} from '../../helpers/auth'

const styles = {
  dashboard: {
    display: 'flex',
    flexDirection: 'column',
  }
};

/*
  Dashboard for students
  Renders a div containing CoursesList, Button for adding course, and a AddCourseModal controlled by button
*/

export default class Dashboard extends Component {

  constructor(props){
    super(props);
    this.closeModal = this.closeModal.bind(this);
    this.handleModal = this.handleModal.bind(this);
    this.state = {
      showModal: false,
    }
  }

  componentDidMount(){
    let courseRef = firebase.database().ref();
    let userUid = getUserUid();
    courseRef.child('users/'+userUid+'/courses').orderByChild('active').equalTo(true).on('value', snap => {
      this.setState({
        courses: snap.val()
      });

    })
  }

  componentWillUnmount(){
    let courseRef = firebase.database().ref();
    let userUid = getUserUid();
    courseRef.child('users/'+userUid+'/courses').orderByChild('active').equalTo(true).off();
  }

  closeModal(){
        this.setState({ showModal: false });
    }


  handleModal(){
        this.setState({
            showModal: true,
        })
    }


  render () {
    return (
    	<div style={styles.dashboard}>
        <h4>My courses</h4>
        <CoursesList courses={this.state.courses}/>
        <Button onClick={this.handleModal} bsStyle="primary"><b>Add new course</b></Button>
        <AddCourseModal showModal={this.state.showModal} closeModal={this.closeModal.bind(this)}/>
      </div>
    )
  }
}