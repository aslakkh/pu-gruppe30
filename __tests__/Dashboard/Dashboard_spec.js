import React from 'react';
import {shallow} from 'enzyme';
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';
import TestUtils from "react-addons-test-utils";
import Dashboard from '../../src/components/Dashboard/Dashboard.js';

const courses = {
    IT5555: {
        active: true,
        goal: 0,
        goals: {},
        oldGoals: {},
        time: 0,
    },
}

describe("Dashboard", function(){

    it('renders', () => {
        shallow(<Dashboard />);
    })

    it('closeModal sets showModal to false', () => {
        const wrapper = shallow(<Dashboard/>);
        wrapper.find('Button').simulate('click');
        
    })
})

test('Dashboard renders student dashboard', () => {
    const component = renderer.create(
        <Dashboard />
    ).toJSON();
    expect(component).toMatchSnapshot(); //render without courses

    const component2 = renderer.create(
        <Dashboard courses={courses} />
    ).toJSON();
    expect(component2).toMatchSnapshot();
})