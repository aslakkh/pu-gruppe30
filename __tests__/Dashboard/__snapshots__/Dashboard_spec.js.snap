// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders student dashboard 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <h4>
    My courses
  </h4>
  <h5>
    You have no active courses. Add a course below.
  </h5>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <b>
      Add new course
    </b>
  </button>
</div>
`;

exports[`Dashboard renders student dashboard 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <h4>
    My courses
  </h4>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        IT5555
      </div>
      <button
        className="btn btn-danger"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
  </ul>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <b>
      Add new course
    </b>
  </button>
</div>
`;
